### ISSUE TITLE
Cancel the query execution if the filter for the widget has been changed

### ISSUE THREAD

**Issue by evshvarov** [intersystems-ru/DeepSeeWeb#84]
How to cancel query execution:
https://community.intersystems.com/post/how-cancel-execution-long-query-deepsee

**Comment by eduard93** [comment]
To do that you need to execute query in Async mode and while it's (maybe) possible to add I think drawbacks from that would outweigh any possible advantage.

**Comment by evshvarov** [comment]
We need it greatly in DSW UI.
Consider the case when you navigate through dashboards, open the Dash1 with several heavy widgets and decided to go back to another Dash2. Now it needs to end ALL the queries in Dash1 which nobody needs. It can take minutes. And only after that, you would be able to get the Dash2 data. In reality, a user already goes to Dash3 (so you even don't need Dash2 widgets) or closes DSW at all. 
So we need it.

**Comment by eduard93** [comment]
Yes, but to do that we need to run all queries asynchronously. It would be slower on the whole because for each query we would need 2 requests at the minimum. 

>And only after that, you would be able to get the Dash2 data.

Why? JS side we just drop the requests. Server has DeepSee agents, so even if one request takes a minute it does not particularly affect other requests.

**Comment by eduard93** [comment]
Added `Timeout` parameter to `/MDX`, if query is not complete before the timeout only basic information is returned to a client. This information includes: `cubeKey`, `queryKey`, `percentDone`.

If `/MDX` is called with `CubeKey` and `QueryKey` parameters instead of `MDX`, then existing query would be returned (or the new `percentDone` would). `Timeout` can be passed to this query too. Default `Timeout` is `0` which means indefinite wait.

Query cancellation at `/CancelQuery/:CubeKey/:QueryKey`.


Commits:
- https://github.com/intersystems-ru/Cache-MDX2JSON/commit/b383736ea09bb92a270223b0d691cba11963f47a
- https://github.com/intersystems-ru/Cache-MDX2JSON/commit/3b5465ad3b87b7087bb1422cff29b320b2f7a7a9

**Comment by evshvarov** [comment]
@gnibeda ? Implement please?

**Comment by Lena-Ev** [comment]
REST cancel all when you go out from the dashboard in session only
@jakcpto

**Comment by Lena-Ev** [comment]
automate requests cancel from dashboards

**Comment by Lena-Ev** [comment]
1. Cancel request when close the tab

2. Cancel request if  we switch between dashboards in one tab

Header -> MDX data exists -> cancel
If te tab g=ahs been opened the header will be there

**Comment by Frisle** [comment]
At this point we achived accuaring query keys for each query, managed deleting the keys when query is done.

We`ve discussed that call for canceling the query should be automatic everytime we received /dashboard. The problem is that each request come one by one and executed accordingly. For query to be canceled request need to be async 

For example on this screenshot marked /MDX request is started on a previous dashboard, but newly opened dashboard have to wait in line for the previos /MDX to load

![Screenshot 2023-10-12 193632](https://github.com/intersystems-community/DeepSeeWeb/assets/47441164/c0b70dac-3ada-430d-b8a5-66d89e44a84b)

We should develop new approach for canceling the queris. I suggest we develop new endpoint for that
